### PR TITLE
Improve encoder decoder model docs

### PR DESCRIPTION
**Pull request by Threepointone4**
# What does this PR do?
This PR improves the documentation of encoder decoder model.

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
       Issues [link](https://github.com/huggingface/transformers/issues/16135)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@patrickvonplaten

### CODE REVIEW

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 1]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
 An application of this architecture could be to leverage two pretrained [`BertModel`] as the encoder
 and decoder for a summarization model as was shown in: [Text Summarization with Pretrained Encoders](https://arxiv.org/abs/1908.08345) by Yang Liu and Mirella Lapata.
 
-The [`~TFEncoderDecoderModel.from_pretrained`] currently doesn't support initializing the model from a
+## Randomely initializing [`EncoderDecoderModel`] from model configurations.
```
Comment (NielsRogge) [inline docs/source/en/model_doc/encoder-decoder.mdx:30]:
```suggestion
## Randomly initializing [`EncoderDecoderModel`] from model configurations.
```

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 2]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+
+[`EncoderDecoderModel`] can be randomly initialized from an encoder and a decoder config. In the following example, we should how to do this using the default [`BertModel`] configuration for both the encoder and the decoder.
```
Comment (NielsRogge) [inline docs/source/en/model_doc/encoder-decoder.mdx:32]:
```suggestion
[`EncoderDecoderModel`] can be randomly initialized from an encoder and a decoder config. In the following example, we show how to do this using the default [`BertModel`] configuration for both the encoder and the decoder.
```

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 3]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+```python
+>>> from transformers import EncoderDecoderModel, BertTokenizer
+
+>>> tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
+>>> model = EncoderDecoderModel.from_encoder_decoder_pretrained("bert-base-uncased", "bert-base-uncased")
+```
+
+## Loading an existing [`EncoderDecoderModel`] checkpoint.
+
+To load fine-tuned checkpoints of the `EncoderDecoderModel` class, ['EncoderDecoderModel`] provides the `from_pretrained(...)` method just like any other model architectures in Transformers.
```
Comment (NielsRogge) [inline docs/source/en/model_doc/encoder-decoder.mdx:60]:
```suggestion
To load fine-tuned checkpoints of the `EncoderDecoderModel` class, ['EncoderDecoderModel`] provides the `from_pretrained(...)` method just like any other model architecture in Transformers.
```

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 4]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
 >>> model = TFEncoderDecoderModel.from_encoder_decoder_pretrained(
 ...     "./encoder", "./decoder", encoder_from_pt=True, decoder_from_pt=True
 ... )
 >>> # This is only for copying some specific attributes of this particular model.
 >>> model.config = _model.config
 ```
 
+## Training
+
+Once the model is created,it can be finetuned simliar to Bart,T5 or any other encoder-decoder model.
```
Comment (NielsRogge) [inline docs/source/en/model_doc/encoder-decoder.mdx:92]:
```suggestion
Once the model is created, it can be fine-tuned simliar to BART, T5 or any other encoder-decoder model.
```

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 5]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+As you can see, only 2 inputs are required for the model in order to compute a loss: `input_ids` (which are the
+`input_ids` of the encoded input sequence) and `labels` (which are the `input_ids` of the encoded
+target sequence).
+
+```python
+>>> from transformers import EncoderDecoderModel, BertTokenizer
```
Comment (NielsRogge) [inline docs/source/en/model_doc/encoder-decoder.mdx:98]:
```suggestion
>>> from transformers import BertTokenizer, EncoderDecoderModel
```
(nit) better to import chronologically

**Review by NielsRogge — APPROVED** [review]
Thanks for improving this! Would be great to also improve the docs of VisionEncoderDecoderModel and SpeechEncoderDecoderModel.

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 6]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+>>> config = EncoderDecoderConfig.from_encoder_decoder_configs(config_encoder, config_decoder)
+>>> model = EncoderDecoderModel(config=config)
+```
+
+## Initialising [`EncoderDecoderModel`] from a pretrained encoder and a pretrained decoder.
+
+[`EncoderDecoderModel`] can be initialized from a pretrained encoder checkpoint and a pretrained decoder checkpoint. Note that any pretrained auto-encoding model, *e.g.* BERT, can serve as the encoder and both pretrained auto-encoding models, *e.g.* BERT, pretrained causal language models, *e.g.* GPT2, as well as the pretrained decoder part of sequence-to-sequence models, *e.g.* decoder of BART, can be used as the decoder.
+Depending on which architecture you choose as the decoder, the cross-attention layers might be randomly initialized.
+Initializing [`EncoderDecoderModel`] from a pretrained encoder and decoder checkpoint requires the model to be fine-tuned on a downstream task, as has been shown in [the *Warm-starting-encoder-decoder blog post*](https://huggingface.co/blog/warm-starting-encoder-decoder).
+To do so, the `EncoderDecoderModel` class provides a [`EncoderDecoderModel.from_encoder_decoder_pretrained`] function.
```
Comment (ydshieh) [inline docs/source/en/model_doc/encoder-decoder.mdx:49]:
```suggestion
To do so, the `EncoderDecoderModel` class provides a [`EncoderDecoderModel.from_encoder_decoder_pretrained`] method.
```

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 7]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+>>> config_decoder = BertConfig()
+
+>>> config = EncoderDecoderConfig.from_encoder_decoder_configs(config_encoder, config_decoder)
+>>> model = EncoderDecoderModel(config=config)
+```
+
+## Initialising [`EncoderDecoderModel`] from a pretrained encoder and a pretrained decoder.
+
+[`EncoderDecoderModel`] can be initialized from a pretrained encoder checkpoint and a pretrained decoder checkpoint. Note that any pretrained auto-encoding model, *e.g.* BERT, can serve as the encoder and both pretrained auto-encoding models, *e.g.* BERT, pretrained causal language models, *e.g.* GPT2, as well as the pretrained decoder part of sequence-to-sequence models, *e.g.* decoder of BART, can be used as the decoder.
+Depending on which architecture you choose as the decoder, the cross-attention layers might be randomly initialized.
```
Comment (ydshieh) [inline docs/source/en/model_doc/encoder-decoder.mdx:47]: I would be more careful to say **the auto-encoding models that provide causal LM implementation**.

Also the sentence is super long, it might be a good idea to split it.

**Review by ydshieh — COMMENTED** [review]
Thank you for the improvement, @Threepointone4 !

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 8]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+
+>>> model.config.decoder_start_token_id = tokenizer.cls_token_id
+>>> model.config.pad_token_id = tokenizer.pad_token_id
+
+>>> input_ids = tokenizer(
+...     "The tower is 324 metres (1,063 ft) tall, about the same height as an 81-storey building, and the tallest structure in Paris. Its base is square, measuring 125 metres (410 ft) on each side.During its construction, the Eiffel Tower surpassed the Washington Monument to become the tallest man-made structure in the world, a title it held for 41 years until the Chrysler Building in New York City was  finished in 1930. It was the first structure to reach a height of 300 metres. Due to the addition of a broadcasting aerial at the top of the tower in 1957, it is now taller than the Chrysler Building by 5.2 metres (17 ft).Excluding transmitters, the Eiffel Tower is the second tallest free-standing structure in France after the Millau Viaduct.",
+...     return_tensors="pt",
+... ).input_ids
+
+>>> labels = tokenizer(
```
Comment (ydshieh) [inline docs/source/en/model_doc/encoder-decoder.mdx:111]: (nit) it might be a good idea to add `>>> # Let's use a summary of the above text as the target`

**Review by ydshieh — APPROVED** [review]

File: docs/source/en/model_doc/encoder-decoder.mdx [hunk 9]
```diff
@@ -25,22 +27,97 @@ any other models (see the examples for more information).
...
+...     "the eiffel tower surpassed the washington monument to become the tallest structure in the world. it was the first structure to reach a height of 300 metres in paris in 1930. it is now taller than the chrysler building by 5. 2 metres ( 17 ft ) and is the second tallest free - standing structure in paris.",
+...     return_tensors="pt",
+... ).input_ids
+
+>>> # the forward function automatically creates the correct decoder_input_ids
+>>> loss = model(input_ids=input_ids, labels=labels).loss
+```
+Detailed [colab](https://colab.research.google.com/drive/1WIk2bxglElfZewOHboPFNj8H44_VAyKE?usp=sharing#scrollTo=ZwQIEhKOrJpl) for training.
```
[the remark this answers is not in the document]
Comment (Threepointone4) [reply docs/source/en/model_doc/encoder-decoder.mdx:119]: @NielsRogge should I remove the Colab link for now ?

**Comment by patrickvonplaten** [comment]
Agree with the reviews of @ydshieh and @NielsRogge ! @Threepointone4 do you want to apply them ? Think we can merge after :-)

File: docs/source/en/model_doc/encoder-decoder.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/en/model_doc/encoder-decoder.mdx:119]: Think ok to leave it for now :-)

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Great job @Threepointone4 ! Merging :-)
